File: day3/part_one.c
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>

#define MAXLINE 1024

int is_symbol(char c)
{
  return (c < '0' || c > '9') && c != '.';
}

int adjacent_to_symbol(char **array, int row, int start_column, int end_column, int rows, int columns)
{
  // top
  for (int i = start_column - 1; i <= end_column + 1; i++)
  {
    if (row - 1 < 0 || i < 0 || i >= columns) continue;
    if (is_symbol(array[row - 1][i]))
    {
      return 1;
    }
  }
  // bottom
  for (int i = start_column - 1; i <= end_column + 1; i++)
  {
    if (row + 1 >= rows || i < 0 || i >= columns) continue;
    if (is_symbol(array[row + 1][i]))
    {
      return 1;
    }
  }
  // left
  int left_idx = start_column - 1;
  if (left_idx >= 0 && is_symbol(array[row][left_idx]))
  {
    return 1;
  }
  // right
  int right_idx = end_column + 1;
  if (right_idx < columns && is_symbol(array[row][right_idx]))
  {
    return 1;
  }

  return 0;
}

int main(int argc, char *argv[])
{
  FILE *fp = fopen("input.txt", "r");
  if (fp == NULL)
  {
    fprintf(stderr, "can't open input.txt\n");
    return 1;
  }

  char buf[MAXLINE];
  int num_lines = 0, row_len = 0;
  while (fgets(buf, MAXLINE, fp) != NULL)
  {
    row_len = strlen(buf);
    num_lines++;
  }

  rewind(fp);

  char **array = malloc(num_lines * sizeof(char*));
  for (int i = 0; i < num_lines; i++)
  {
    array[i] = malloc(row_len * sizeof(char));
  }

  int line = 0;
  while (fgets(buf, MAXLINE, fp) != NULL)
  {
    buf[strcspn(buf, "\n")] = 0;
    strcpy(array[line++], buf);
  }

  fclose(fp);
  
  int ret = 0;
  for(int i = 0; i < num_lines; i++)
  {
    int start_column = -1, end_column = -1;
    for (int j = 0; j < row_len; j++)
    {
      if (isdigit((unsigned char)array[i][j]))
      {
        if (start_column == -1)
        {
          start_column = j;
        }
        end_column = j;
      }
      else if (start_column != -1 && end_column != -1)
      {
        if(adjacent_to_symbol(array, i, start_column, end_column, num_lines, row_len - 1))
        {
          int value = 0;
          sscanf(&array[i][start_column], "%d", &value);
          ret += value;
        }
        start_column = end_column = -1;
      }
    }
  }

  free(array);
  
  printf("Answer %d", ret);
  return 0;
}

File: day10/part_one.py
from sys import setrecursionlimit

def compute_circle_len():
    all_lines = []
    with open("input.txt", "r") as file:
        for line in file.readlines():
            all_lines.append(list(line.strip()))
    
    # find S
    s_position = []
    for i in range(len(all_lines)):
        if "S" not in all_lines[i]:
            continue
        s_position = [i, all_lines[i].index("S")]

    ret = 0

    def dfs(i, row, col):
        nonlocal ret

        if row < 0 or col < 0:
            return None
        if row >= len(all_lines) or col >= len(all_lines[0]):
            return None

        all_lines[row][col] = 1
        ret = max(ret, i)

        # top
        if row > 0 and all_lines[row - 1][col] in ['|', 'F', '7']:
            dfs(i + 1, row - 1, col)

        # bottom
        if row < len(all_lines) - 1 and all_lines[row + 1][col] in ['|', 'J', 'L']:
            dfs(i + 1, row + 1, col)

        # left
        if col > 0 and all_lines[row][col - 1] in ['-', 'F', 'L']:
            dfs(i + 1, row, col - 1)

        # right
        if col < len(all_lines[0]) - 1 and all_lines[row][col + 1] in ['-', '7', 'J']:
            dfs(i + 1, row, col + 1)

    setrecursionlimit(30000)
    dfs(0, s_position[0], s_position[1])
    print("Answer:", (ret + 1) / 2)

compute_circle_len()
